File: backend/python-flask/app/routes/itineraries.py
from flask import Blueprint, jsonify, request
from app.db import db
from app.models.itinerary import Itinerary, ItineraryStop

itineraries_bp = Blueprint('itineraries', __name__)

# ============================================================
#  Itinerary Routes — Pre-built (no implementation needed)
#
#  These endpoints handle saving and retrieving trip itineraries.
# ============================================================


# POST /api/itineraries — Save an optimised route
@itineraries_bp.route('', methods=['POST'])
def create():
    data = request.get_json()

    itinerary = Itinerary(
        strategy=data['strategy'],
        total_distance=data['totalDistance'],
    )
    db.session.add(itinerary)
    db.session.flush()  # Get the ID before committing

    for stop in data['stops']:
        itinerary_stop = ItineraryStop(
            itinerary_id=itinerary.id,
            stop_number=stop['stopNumber'],
            match_id=stop['match']['id'],
            city_id=stop['city']['id'],
            distance_from_previous=stop.get('distanceFromPrevious', 0),
        )
        db.session.add(itinerary_stop)

    db.session.commit()
    return jsonify({'id': itinerary.id}), 201


# GET /api/itineraries/<id> — Retrieve a saved itinerary
@itineraries_bp.route('/<id>', methods=['GET'])
def get_by_id(id):
    itinerary = Itinerary.query.get(id)

    if not itinerary:
        return jsonify({'error': 'Itinerary not found'}), 404

    return jsonify(itinerary.to_dict()), 200

File: backend/node-express/src/routes/itineraries.ts
import { Router } from 'express';
import * as ItineraryModel from '../models/Itinerary';

const router = Router();

/**
 * Itinerary Routes — Pre-built (no implementation needed)
 *
 * These endpoints handle saving and retrieving trip itineraries.
 */

// POST /api/itineraries — Save an optimised route
router.post('/', (req, res) => {
  try {
    const route = req.body;
    const result = ItineraryModel.create(route);
    res.status(201).json(result);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create itinerary' });
  }
});

// GET /api/itineraries/:id — Retrieve a saved itinerary
router.get('/:id', (req, res) => {
  try {
    const { id } = req.params;
    const itinerary = ItineraryModel.getById(id);

    if (!itinerary) {
      res.status(404).json({ error: 'Itinerary not found' });
      return;
    }

    res.json(itinerary);
  } catch (error) {
    res.status(500).json({ error: 'Failed to retrieve itinerary' });
  }
});

export default router;
